include/nbl/builtin/hlsl/emulated/matrix_t.hlsl

```diff
@@ -0,0 +1,76 @@
+#ifndef _NBL_BUILTIN_HLSL_EMULATED_MATRIX_T_HLSL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_EMULATED_MATRIX_T_HLSL_INCLUDED_
+
+#include <nbl/builtin/hlsl/portable/float64_t.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+
+template<typename EmulatedType, uint32_t N, uint32_t M>
+struct emulated_matrix {};
+
+template<typename EmulatedType>
+struct emulated_matrix<EmulatedType, 2, 2>
+{
+    using vec_t = emulated_vector_t2<EmulatedType>;
+    using this_t = emulated_matrix<EmulatedType, 2, 2>;
+
+    vec_t rows[2];
+
+    this_t getTransposed() NBL_CONST_MEMBER_FUNC
+    {
+        this_t output;
+
+        output.rows[0].x = rows[0].x;
+        output.rows[1].x = rows[0].y;
+
+        output.rows[0].y = rows[1].x;
+        output.rows[1].y = rows[1].y;
+
+        return output;
+    }
+};
+
+template<typename EmulatedType>
+struct emulated_matrix<EmulatedType, 3, 3>
+{
+    using vec_t = emulated_vector_t3<EmulatedType>;
+    using this_t = emulated_matrix<EmulatedType, 3, 3>;
+
+    vec_t rows[3];
+
+    this_t getTransposed() NBL_CONST_MEMBER_FUNC
+    {
+        this_t output;
+
+        output.rows[0].x = rows[0].x;
+        output.rows[1].x = rows[0].y;
+        output.rows[2].x = rows[0].z;
+
+        output.rows[0].y = rows[1].x;
+        output.rows[1].y = rows[1].y;
+        output.rows[2].y = rows[1].z;
+
+        output.rows[0].z = rows[2].x;
+        output.rows[1].z = rows[2].y;
+        output.rows[2].z = rows[2].z;
+
+        return output;
+    }
+
+    vec_t operator[](uint32_t rowIdx)
+    {
+        return rows[rowIdx];
+    }
+};
```

Comment:
why not declare a 
```cpp
namespace nbl::hlsl
{

namespace impl
{
template<template<typename,uint32_t,uint32_t> Matrix, typename T, uint32_t N, uint32_t M>
struct transpose_helper
{
   Matrix<T,M,N> operator()(NBL_CONST_REF(Matrix<T,N,M>) m)
   {
       return ::transpose(m);
   }
};
}

template<template<typename,uint32_t,uint32_t> Matrix, typename T, uint32_t N, uint32_t M>
Matrix<T,M,N> operator(NBL_CONST_REF(Matrix<T,N,M>) m)
{
   const impl::tranpose_helper<Matrix,T,N,M> fn;
   return fn(m);
}

}
```

and then partial spec for the `emulated_matrix` ?


Then you wouldn't need to partial spec the `emulated_matrix` structs (of which there are like 16 or 9 specs)

Reply:
btw the native HLSL type might require us to use `int32_t` instead of `uint32_t` for the template params

Reply:
regarding the `transpose_helper` struct you suggested, I think i've figured out a better solution:
https://github.com/Devsh-Graphics-Programming/Nabla/blob/emulated_float64_t/include/nbl/builtin/hlsl/emulated/matrix_t.hlsl#L23-L36

Reply:
> btw the native HLSL type might require us to use `int32_t` instead of `uint32_t` for the template params

yes, later I will extended `emulated_float_tests` example to test emulated matrices and vectors too

Reply:
> regarding the `transpose_helper` struct you suggested, I think i've figured out a better solution: https://github.com/Devsh-Graphics-Programming/Nabla/blob/emulated_float64_t/include/nbl/builtin/hlsl/emulated/matrix_t.hlsl#L23-L36

neat, but we still need global `nbl::hlsl::mul` and `nbl::hlsl::transpose` so that we can write code like
```cpp
mul(M,V);
tranpose(M);
```
without caring if `M` is `::matrix` or `nbl::hlsl::emulated_matrix_t`

much how our `static_cast` works